git.yazi/main.lua

```diff
@@ -48,119 +107,132 @@ local function root(cwd)
 	until not cwd
 end
 
-local function bubble_up(changed)
+---@param changes Changes
+---@return Changes
+local function bubble_up(changes)
 	local new, empty = {}, Url("")
-	for k, v in pairs(changed) do
-		if v ~= 3 and v ~= 30 then
-			local url = Url(k):parent()
-			while url and url ~= empty do
-				local s = tostring(url)
-				new[s] = (new[s] or 0) > v and new[s] or v
-				url = url:parent()
-			end
+	for path, status in pairs(changes) do
+		local url = Url(path):parent()
+		while url and url ~= empty do
+			local s = tostring(url)
+			new[s] = (new[s] or Status.unkonwn) > status and new[s] or status
+			url = url:parent()
 		end
 	end
 	return new
 end
 
+---@param ignored Changes
+---@param cwd Url
+---@param repo Url
 local function propagate_down(ignored, cwd, repo)
 	local new, rel = {}, cwd:strip_prefix(repo)
-	for k, v in pairs(ignored) do
-		if v == 30 then
-			if rel:starts_with(k) then
-				new[tostring(repo:join(rel))] = 30
-			elseif cwd == repo:join(k):parent() then
-				new[k] = 3
-			end
+	for _, path in ipairs(ignored) do
+		if rel:starts_with(path) then
+			-- if `cwd` is a subfolder of an ignored directory, mark the `cwd` as `__ignored_dir__`
+			new[tostring(cwd)] = "__ignored_dir__"
```

Comment:
string is not an immediate value, which means that transferring them between threads requires constant memory reallocation, and it also actually occupies more memory than just a number.